src/StackDeploy/CLI.hs

```diff
@@ -0,0 +1,145 @@
+
+import Options.Applicative
+  ( CommandFields
+  , Mod
+  , Parser
+  , ParserInfo
+  , ReadM
+  , argument
+  , command
+  , eitherReader
+  , help
+  , helper
+  , hsubparser
+  , idm
+  , info
+  , long
+  , metavar
+  , option
+  , str
+  )
+
+type InstanceSpecProvider = forall m . MonadIO m
+  => (Name -> [Parameter]
+  -> m InstanceSpec)
+
+parserInfo :: forall m r . (AWSConstraint r m, MonadAWS m)
+           => InstanceSpecProvider
+           -> ParserInfo (m ExitCode)
+parserInfo provider = wrapHelper commands
+  where
+    commands :: Parser (m ExitCode)
+    commands = hsubparser
+      $  mkCommand "delete" (delete <$> nameParser)
+      <> mkCommand "events" (events <$> nameParser)
+      <> mkCommand "list"   (pure list)
+      <> mkCommand "sync"   (sync <$> nameParser <*> many parameterParser)
+      <> mkCommand "token"  (pure printNewToken)
+      <> mkCommand "wait"   (wait <$> nameParser <*> tokenParser)
+
+    nameParser :: Parser Name
+    nameParser = Name <$> argument str (metavar "NAME")
+
+    tokenParser :: Parser Token
+    tokenParser = Token <$> argument str (metavar "TOKEN")
+
+    mkCommand :: String -> Parser b -> Mod CommandFields b
+    mkCommand name = command name . wrapHelper
+
+    wrapHelper :: Parser b -> ParserInfo b
+    wrapHelper parser = info (helper <*> parser) idm
+
+    sync :: Name -> [Parameter] -> m ExitCode
+    sync name params = do
+      spec <- provider name params
+      exitCode
+        =<< perform . maybe (OpCreate name spec) (`OpUpdate` spec)
+        =<< getStack name
+
+    wait :: Name -> Token -> m ExitCode
+    wait name token = maybe success (waitForOperation token) =<< getStack name
+
+    delete :: Name -> m ExitCode
+    delete = maybe success (exitCode <=< perform . OpDelete) <=< getStack
+
+    list :: m ExitCode
+    list = do
+      runConduit $ stackNames .| mapM_ say
+      success
+
+    events :: Name -> m ExitCode
+    events name = do
+      runConduit $ listResource req dsersStackEvents .| mapM_ printEvent
+      success
+      where
+        req = describeStackEvents & dseStackName .~ pure (toText name)
+
+    waitForOperation :: Token -> Id -> m ExitCode
+    waitForOperation token stackId = do
+      sayStack stackId "wait"
+      status <- waitForAccept RemoteOperation{..} printEvent
+      sayStack stackId $ finalMessage status
+      exitCode status
+
+    printNewToken :: m ExitCode
+    printNewToken = do
+      say =<< newToken
+      success
+
+    success :: m ExitCode
+    success = pure ExitSuccess
+
+    exitCode = \case
+      RemoteOperationSuccess -> success
+      RemoteOperationFailure -> pure $ ExitFailure 1
+
+parameterParser :: Parser Parameter
+parameterParser = option
+  parameterReader
+  (long "parameter" <> help "Set stack parameter")
+
+parameterReader :: ReadM Parameter
+parameterReader = eitherReader (Text.parseOnly parser . convertText)
+  where
+    parser = do
+      key <- convertText <$> Text.many1 (Text.satisfy allowChar)
```

Comment:
Do the keys have any more specific format? Such as "must always begin with capital letter"?

I realize we probably want to defer this, but it has been proven that such constraints when added early on have disproportionate benefits.

Reply:
There are more rules, for now I just do the minimum that works.

I intend to import a full blown most restrictive parser I made for another project. But not at this point.